---
title: "SI File 3: Modelling Gene Expression with Inducer Half-Live"
author: "Rainer Machne"
date: "Oct 07, 2021"
bibliography: genex.bib
output: 
    pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Assuming that the inducer half-life $t_{1/2}=\frac{\log 2}{\delta}$ is not affected by metabolism, and that the measured fluorescence per OD is linearly related to protein concentration, $f_\text{OD}\sim y(t)$, we can describe the measured data as:

\begin{equation}
\label{eq:de}
\begin{aligned}
\beta &= \mu + \delta_P\\
I(t) &= I_0 e^{-\delta t}\\
\alpha(t) &= \ell + v \frac{I^n}{I^n+K^n}\\
\frac{\text{d}y}{\text{d} t} = y' &= \alpha - \beta y
\end{aligned}
\end{equation}

with culture growth rate $\mu$ and AVS-tagged protein degradation rate 
$\delta_P$, basal ("leaky") transcription rate $\ell$, maximal induced
transcription rate $v$, half-maximal inducer concentration $K$ and a
Hill-factor $n$, and  initial inducer concentration $I_0$ and inducer
degradation  rate $\delta$.

Note, that the Michaels-Menten-type activation function $\alpha$ is
empirical, aimed at fitting the data at hand. The model simplifies
multiple steps of complex formation (inducer and transcription factor,
transcription factor on promoter) and subsumes both transcription and
translation. Nevertheless, the parameters may roughly reflect dominant
biochemical parameters of the real system, eg. $K$ could be
representative of the actual dissociation constant between inducer and
transcription factor $K=\frac{I P}{P}$, or $n>1$ may reflect multiple
binding sites of the complex on the promoter.

This inhomogenous first-order differential equation can be solved 
for a constant inducer, $\delta=0$, (Appendix I):

\begin{equation}
\label{eq:consti}
y(t) = y_0 e^{-\beta t} +\frac{1- e^{-\beta t}}{\beta} \left(\ell + v \frac{I^n}{I^n+K^n}\right)\,.
\end{equation}

An analytic solution for the general case, $\delta >0$, is harder
(Appendix II).  The integration routine of the
[Wolfram alpha web service](https://www.wolframalpha.com/input/)
provides the solution which simply replaces the activation term with
the hypergeometric function $_2\text{F}_1$:

\begin{equation}
\label{eq:vari}
y(t) = y_0 e^{-\beta t} + \frac{1- e^{-\beta t}}{\beta}\left(\ell + v\; _2\text{F}_1\left(1,\frac{\beta}{n \delta};\frac{\beta}{n \delta}+1;-e^{n\delta t}\frac{K^n}{I_0^n}\right)\right)\,.
\end{equation}

$_2\text{F}_1$ is implemented in the GNU Scientific Library (GSL), but
requires certain transformations of the variables [@Forrey1997] to
ensure convergence (Appendix III). For our model $_2\text{F}_1$ is well
defined only for $\delta>0$ and $I_0>0$ and we use
eq.(\ref{eq:consti}) for $I_0=0$ or $\delta=0$ (no or constant
inducer) and eq.(\ref{eq:vari}) for $\delta>0$, to fit experimental
data.

We observed numerical stability problems for too small $\delta$,
esp. when combined with small $I_0$. A more stable implementation of
$_2\text{F}_1$ or the required transformations would be
desirable. Alternatively, a scaling to relative time $t/\delta$ may
avoid the problematic terms. However, the approach worked well for the
observed inducer half-lives.



## Implementation & Example

The model was implemented in R, including a wrapper for required data
transformations for the $_2\text{F}_1$ function, which is available
via the R package `gsl`. The R code is freely available at
https://github.com/raim/genex. The following code shows how to use
the implemented functions, using typical parameters for induced GFP
expression in *Escherichia coli* measured over all growth phases in
microtiter 96-well plate cultures, and assuming an inducer half life
$\tau_{1/2}=\frac{\ln 2}{\delta}=$ 17.5 h, within the ranges
of half-lives (10--40 h) measured for anhydro-tetracycline by @Politi2014.



```{r functions, echo=TRUE, warning=FALSE}
source("../R/genex.R")
### GENERAL EXPERIMENT PARAMETERS 
## experiment time
time <- seq(0,200,5) # h

##	INITIAL INDUCER CONCENTRATIONS 
i0 <- c(10000, 100) # for time series plot
I0 <- c(0,10^seq(-1,4,.1)) # a.u., for dose-response plot
ts <- c(24,48)      # sampling time points for dose-response

## growth and protein degradation rates
mu <- 0.029 # h-1, 1 division per day - mu = log(2)/24
dy <- 0.07  # h-1, protein degradation rate, ASV-tagged mVenus
beta <- mu + dy

### INDUCER-SPECIFIC PARAMETERS
y0 <- 50 # intial protein concentration
n <- 2   # Hill factor, ~ number of binding sites 
K <- 100 # half-maximal concentration 
l <- 1   # uninduced "leaky" transcription
v <- 100 # maximal induced transcription

## INDUCER HALF-LIFE - constant inducer over time
tI <- 17.5 # h, half-life aTc
delta <- log(2)/tI # h-1, degradation rate aTc
```

```{r solve, echo=TRUE, warning=FALSE}
## calculate time series and dose-response curves using the `fexpr`
## function provided in R/genex.R

## CALCULATE TIME SERIES
yt1 <- fexpr(time=time, I0=i0[1], delta=delta, beta=beta, 
             y0=y0, n=n, K=K, l=l, v=v)
yt2 <- fexpr(time=time, I0=i0[2], delta=delta, beta=beta, 
             y0=y0, n=n, K=K, l=l, v=v)
## CALCULATE DOSE-RESPONSE CURVES			 
yi1 <- fexpr(time=ts[2], I0=I0, delta=delta, beta=beta, 
             y0=y0, n=n, K=K, l=l, v=v)
yi2 <- fexpr(time=ts[1], I0=I0, delta=delta, beta=beta, 
             y0=y0, n=n, K=K, l=l, v=v)

```

```{r plot, echo=FALSE, warning=FALSE}
ys <- c(yt1,yt2,yi1,yi2) # get common ylim for plots
ys <- ys[is.finite(ys)]
ylim <- c(0, max(ys,na.rm=TRUE))

par(mfcol=c(2,1), mai=c(.5,.5,.1,.1), mgp=c(1.3,.4,0),tcl=-.25)
plot(time, yt1, type="b", pch=19, cex=.5, ylim=ylim,
     ylab="fluorescence/OD, a.u.", xlab="time, h")
lines(time, yt2, col=2, type="b", pch=19, cex=.5)
legend("topright",legend=c(as.expression(bquote(I[0] ~":"~.(i0[1]))),
                           bquote(I[0]~":"~.(i0[2]))),
       title=paste0("half-life: ",tI, " h"), col=1:2,lty=1,pch=19,pt.cex=.5)
abline(v=ts, col=3:4, lty=1)
plot(I0, yi1, type="b", col=4, lty=1, pch=19, cex=.5,
     log="x",ylim=ylim,xlim=c(1,max(I0)),
     xlab="inducer concentration", ylab="fluorescence/OD, au")
lines(I0, yi2, type="b", col=3, lty=1, pch=19, cex=.5)
abline(v=i0,col=1:2)
legend("bottomright",paste("at",ts,"h"),
       title=paste0("half-life: ",tI, " h"),
       col=3:4,lty=1,pch=19,pt.cex=.5, bg="white")
```

We first calculate (eq. \ref{eq:vari}) protein concentration time
courses for two different inducer concentrations (upper panel).  Next,
we scan over inducer concentrations to generate dose-response curves,
sampled at two different time points (lower panel).  In the upper
panel, vertical lines (green and blue) indicate the sampling time
points used for the two dose-response curves in the lower panel. In
the lower panel, vertical lines (red and black) indicate the
concentrations used for the two time series in the upper panel.

**This shows how the measured dose response depends on the sampling
time**.  At low inducer concentration the signal (protein
concentration) is higher when sampled at 24 h than at 48 h. At higher
inducer concentration the maximal signal is seen later, and higher at
48 h than at 24 h.

## Parameter Fitting Strategy and Results

We roughly followed the following strategy to fit the various data
sets:

1. Calculate growth rate $\mu$ for all experiments,
2. Calculate $\mu$ and $\beta =\mu + \delta_P$ from rhamnose experiment
after step-down to get ASV-tagged protein degradation rate $\delta_P$,
3. Estimate $y_0$ from 0 inducer for each promoter.
4. Fit $v$ and  $K$ from dose-response curves using a rough estimate of inducer
half-life $\delta$,
5. Use obtained parameters as initial values and fit half-lives $\delta$ from
time-series experiments,
6. Iterate the above procedure until all shared parameters converge.


\newpage
# APPENDIX

## Appendix I: Induced Transcription with Constant Inducer

The "inhomogenous linear first order differential equation", 
$$
y'(t) + \beta y(t) = \alpha\,,
$$ 
can be solved by adding a particular solution $y_p$ of the inhomogenous
equation and the general solution $y_h$ of the homogenous version of the equation 
(with $\alpha=0$).

The general solution is $y_h = C e^{-\beta t}$, and the particular solution
is simply $y_p=\frac{\alpha}{\beta}$, and we get:

$$
y(t) = y_p+y_h = \frac{\alpha}{\beta} + C e^{-\beta t}\,.
$$


Assuming an initial condition $y(0)=y_0$ we further get $C=y_0-\frac{\alpha}{\beta}$, and

$$
y(t) = y_0 e^{-\beta t} + \frac{\alpha}{\beta} \left(1 - e^{-\beta t}\right)\,.
$$

## Appendix II: Induced Transcription with Inducer Half-Life

Here the factor $\alpha$ itself is a function of time $t$,
and can use the "Variation of the Constant" method, where the
constant $C$ of the solution of the homogenous equation is replaced
by a function $f(t)$, ie.

\begin{equation}
\label{eq:voc}
y(t) = f(t)\,e^{-\beta t}
\end{equation}

Using this in the full equation, we get:

$$
f'(t) e^{-\beta t} + f(t) e^{-\beta t} (-\beta) + \beta f(t) e^{-\beta t} = \alpha(t)\,
$$
where two terms on the left-hand side cancel out, and we need to integrate to get $f(t)$:

$$
f(t) = \int \alpha(t) e^{\beta t} \text{d}t + C 
$$

Let's first make it simple, and assume a linear induction with $I\ll K$, with

$$
\alpha(t) \approx v \frac{I_0 e^{-\delta t}}{K}\,.
$$
This is comparatively easy to integrate and yields

$$
f(t) = \frac{v}{\beta - \delta} \frac{I_0}{K} e^{(\beta-\delta)t} + C\,.
$$

Again solving for $y(0)=y_0$ we obtain an expression for $C$
and get:

$$
y(t) = y_0 e^{-\beta t} + \frac{v}{\beta - \delta} \frac{I_0}{K} \left( e^{-\delta t} - e^{-\beta t}\right)\,.
$$

Next, we attempt to integrate the more complex activation function:

$$
\begin{aligned}
\alpha(t) =& \ell + v \frac{(I_0 e^{-\delta t})^n}{(I_0 e^{-\delta t})^n + K^n}\\
f(t) =& \int \alpha(t) e^{\beta t} \text{d}t + C
\end{aligned}
$$

At [Wolfram Alpha](https://www.wolframalpha.com/) (replacing $I_0$ by `A` and $t$ by `x`)\newline
`integrate ( (l + v*A^n*e^(-d*n*x)/(A^n*e^(-d*n*x) + K^n))*e^(b*x))`
yields\newline
`(e^(b x) (v 2F1(1, b/(d n), b/(d n) + 1, -A^(-n) e^(d n x) K^n) + l))/b + constant`,\newline
where $_2\text{F}_1$ is the hypergeometric function, ie.:

\begin{equation}
\label{eq:wasln}
f(t) = \frac{e^{\beta t}}{\beta}\left( _2\text{F}_1\left(1,\frac{\beta}{n \delta};\frac{\beta}{n \delta}+1;-e^{n\delta t}\frac{K^n}{I_0^n}\right)v + l \right) + C\,.
\end{equation}

We could not identify the applied integration rule in literature, but
minimally, for $\beta\ne 1$: $\int \frac{e^{\beta t}}{1+e^t}
\text{d}t= \frac{e^{\beta t}}{\beta}\,
_2\text{F}_1(1,\beta;\beta+1;-e^t)$).

Substituting eq. \ref{eq:wasln} in eq. \ref{eq:voc} and solving for $y(0)=y_0$, we get:

$$
y(t) = y_0 e^{-\beta t} + \frac{1}{\beta}\left( _2\text{F}_1\left(1,\frac{\beta}{n \delta};\frac{\beta}{n \delta}+1;-e^{n\delta t}\frac{K^n}{I_0^n}\right)v + l \right) (1- e^{-\beta t})\,.
$$

## Appendix III: The Hypergeometric Function

The hypergeometric function $_2\text{F}_1$ is rather unusual but often turns up in
solutions of differential equations. See eg. Daalhuis' presentation at NIST
(https://dlmf.nist.gov/15.17) for more information.
It can be solved using the `R`  interface to its implementation
in the "GNU Scientific Library", specifically the function `gsl::hyperg_2F1`. 
However, (fast) convergence is only guaranteed for $|z|<1$, and may fail for certain
combinations of the other parameters. For our model $z<0$ is always true, but
$|z|$ can become very large, eg. it is exponential with $t$, and inversely
proportional to $I_0$. We use the transformation provided by @Forrey1997
for $z< -1$:

```{r, eval=FALSE}
library(gsl)
Gauss2F1_forrey <- function(a,b,c,z){
  if( abs(z) < 1 ) {
    y <- hyperg_2F1(a, b, c, z)
  } else {
    w <- 1/(1-z)
    A <- w^a*gamma(c)*gamma(b-a)/(gamma(b)*gamma(c-a))*hyperg_2F1(a,c-b,a-b+1,w)
    B <- w^b*gamma(c)*gamma(a-b)/(gamma(a)*gamma(c-b))*hyperg_2F1(b,c-a,b-a+1,w)
    y <- A+B
  }
  y
}
```

An alternative approach was suggested at `stackexchange`
by Stephane Laurent which seems more efficient. However, no official reference
for this solution was provided. 

```{r, eval=FALSE}
## https://stats.stackexchange.com/questions/33451/computation-of-hypergeometric-function-in-r
## quote: "Based on my experience I also recommend to only evaluate the Gauss 
## hypergeometric function for a value of the variable lying in [0,1], and 
## use a transformation formula for values in ]-Inf,0]."
## Alternatively, use a more precise implementation at
## http://stla.overblog.com/the-binary-splitting-with-the-r-gmp-package-application-to-gauss-hypergeometric-function
library(gsl)
Gauss2F1_laurent <- function(a,b,c,z){
  if(z>=0 & z<1){
    hyperg_2F1(a,b,c,z)
  }else{
    hyperg_2F1(c-a,b,c,1-1/(1-z))/(1-z)^b
  }
}
```

# References
<div id="refs"></div>


